src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,459 @@
+                                                                                      isPortedFxCopRule: false,
+                                                                                      isDataflowRule: false);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(MinimumOsRule, ObsoleteRule, RemovedRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var typeName = WellKnownTypeNames.SystemRuntimeInteropServicesRuntimeInformation;
+
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(typeName + "Helper", out var runtimeInformationType))
+                {
+                    runtimeInformationType = context.Compilation.GetOrCreateTypeByMetadataName(typeName);
+                }
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesOSPlatform, out var osPlatformType))
+                {
+                    return;
+                }
+
+                var guardMethods = GetRuntimePlatformGuardMethods(runtimeInformationType!);
+
+                context.RegisterOperationBlockStartAction(context => AnalyzerOperationBlock(context, guardMethods, osPlatformType));
+            });
+
+            static ImmutableArray<IMethodSymbol> GetRuntimePlatformGuardMethods(INamedTypeSymbol runtimeInformationType)
+            {
+                return runtimeInformationType.GetMembers().OfType<IMethodSymbol>().Where(m =>
+                    s_platformCheckMethodNames.Contains(m.Name) && !m.Parameters.IsEmpty).ToImmutableArray();
+            }
+        }
+
+        private void AnalyzerOperationBlock(OperationBlockStartAnalysisContext context, ImmutableArray<IMethodSymbol> guardMethods, INamedTypeSymbol osPlatformType)
+        {
+            var parsedTfms = ParseTfm(context.Options, context.Compilation, context.CancellationToken);
+            var platformSpecificOperations = PooledDictionary<IOperation, ImmutableArray<PlatformAttributeInfo>>.GetInstance();
+
+            context.RegisterOperationAction(context =>
+            {
+                AnalyzeInvocationOperation(context.Operation, context, parsedTfms, ref platformSpecificOperations);
+            }
+            , OperationKind.Invocation, OperationKind.PropertyReference, OperationKind.FieldReference);
+
+            context.RegisterOperationBlockEndAction(context =>
+            {
+                try
+                {
+                    if (platformSpecificOperations.Count == 0)
+                    {
+                        return;
+                    }
+
+                    if (guardMethods.IsEmpty || !(context.OperationBlocks.GetControlFlowGraph() is { } cfg))
```

Comment:
guardMethods are new API so if it does not exist in the current compilation no need to do flow analysis, just report diagnostic for all methods in platformSpecificOperations dictionary

Reply:
> guardMethods are new API so if it does not exist in the current compilation

Right, but we could have bailed much earlier, no?

Reply:
Don't we want to warn on builds not having those methods?